### ISSUE TITLE
CellariumGPT refactoring

### ISSUE THREAD

**Issue by mbabadi** [cellarium-ai/cellarium-ml#131]
- [ ] Refactoring tokenization (discussion below)

> Can you also separately take obs_total_mrna_umis_n (to be concatenated to values_nc) and out_total_mrna_umis_n to generate a token specifying the library size we need the readout to be at? This construct can be used for model training with downsampling. We could use Brice's logic to Duplicate the counts, downsample the first copy to use for observation, and use the second copy for readout.
>
> Another thought is to have multiple tokenize methods. For example:
-- generate_observation_tokens
-- generate_output_tokens
-- generate_register_tokens (later)
-- generate_metadata_tokens (later)
>
> In this construct, generate_observation_tokens takes x_ng, obs_total_mrna_umis_n as input and generates a bunch of tokens. generate_output_tokens, for now, just takes out_total_mrna_umis_n and generates a single token.
>
> All of the tokens are then concatenated and given to an embedding layer. The tokens should also be equipped with metadata to specify how they should be embedded ...

- [ ] CellariumBERT "denoiser" model: duplicating the data, applying noise on one copy (e.g. binomial downsampling, binomial upsampling, dropout), using the noisy copy as observation tokens, and used the master copy for loss. Attention must be paid to ensure that needed summary stats for tokenization (e.g. total_mrna_umis_n) are calculated from the appropriate copy.
 
- [ ] Different training strategies with maximal code reuse.  One idea is to have different models CellariumBERT, CellariumDecoderOnlyGPT, CellariumEncoderDecoderGPT, etc. but with shared infrastructure (transformer base, tokenization methods, etc).

- [ ] (more difficult) Flexibility to take the transformer backbone from a checkpoint (e.g. from CellariumBERT) and use it to pretrain another model (e.g. CellariumDecoderOnlyGPT).
